File: problems/40_combination_sum_ii/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), res);
        return res;
    }
    private void backtrack(int[] cands, int remain, int start, List<Integer> path, List<List<Integer>> res) {
        if (remain == 0) { res.add(new ArrayList<>(path)); return; }
        for (int i = start; i < cands.length; i++) {
            if (cands[i] > remain) break;
            if (i > start && cands[i] == cands[i - 1]) continue;
            path.add(cands[i]);
            backtrack(cands, remain - cands[i], i + 1, path, res);
            path.remove(path.size() - 1);
        }
    }
}

File: problems/40_combination_sum_ii/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;
        function<void(int, int)> bt = [&](int start, int remain) {
            if (remain == 0) { res.push_back(path); return; }
            for (int i = start; i < candidates.size(); i++) {
                if (candidates[i] > remain) break;
                if (i > start && candidates[i] == candidates[i - 1]) continue;
                path.push_back(candidates[i]);
                bt(i + 1, remain - candidates[i]);
                path.pop_back();
            }
        };
        bt(0, target);
        return res;
    }
};
